crates/trippy-core/src/tos.rs

```diff
@@ -0,0 +1,206 @@
+        }
+    }
+}
+
+impl Display for ExplicitCongestionNotification {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::NotECT => write!(f, "Not ECN-Capable Transport"),
+            Self::ECT1 => write!(f, "ECN Capable Transport(1)"),
+            Self::ECT0 => write!(f, "ECN Capable Transport(0)"),
+            Self::CE => write!(f, "Congestion Experienced"),
+        }
+    }
+}
+
+/// Differentiated Services Code Point (`DSCP`).
+///
+/// This is used in the `DSCP` field of the `IP` header.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum DifferentiatedServicesCodePoint {
+    /// Best Effort (000 000, 0 dec)
+    BE,
+    /// Assured Forwarding 11 (001 010, 10 dec)
+    AF11,
+    /// Assured Forwarding 12 (001 100, 12 dec)
+    AF12,
+    /// Assured Forwarding 13 (001 110, 14 dec)
+    AF13,
+    /// Assured Forwarding 21 (010 010, 18 dec)
+    AF21,
+    /// Assured Forwarding 22 (010 100, 20 dec)
+    AF22,
+    /// Assured Forwarding 23 (010 110, 22 dec)
+    AF23,
+    /// Assured Forwarding 31 (011 010, 26 dec)
+    AF31,
+    /// Assured Forwarding 32 (011 100, 28 dec)
+    AF32,
+    /// Assured Forwarding 33 (011 110, 30 dec)
+    AF33,
+    /// Assured Forwarding 41 (100 010, 34 dec)
+    AF41,
+    /// Assured Forwarding 42 (100 100, 36 dec)
+    AF42,
+    /// Assured Forwarding 43 (100 110, 38 dec)
+    AF43,
+    /// Class Selector 1 (001 000, 8 dec)
+    CS1,
+    /// Class Selector 2 (010 000, 16 dec)
+    CS2,
+    /// Class Selector 3 (011 000, 24 dec)
+    CS3,
+    /// Class Selector 4 (100 000, 32 dec)
+    CS4,
+    /// Class Selector 5 (101 000, 40 dec)
+    CS5,
+    /// Class Selector 6 (110 000, 48 dec)
+    CS6,
+    /// Class Selector 7 (111 000, 56 dec)
+    CS7,
+    /// High Priority Expedited Forwarding (101 110, 46 dec)
+    EF,
+    /// Other DSCP value (not defined in the standard)
+    Other(u8),
+}
+
+impl TryFrom<u8> for DifferentiatedServicesCodePoint {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(Self::BE),
+            10 => Ok(Self::AF11),
+            12 => Ok(Self::AF12),
+            14 => Ok(Self::AF13),
+            18 => Ok(Self::AF21),
+            20 => Ok(Self::AF22),
+            22 => Ok(Self::AF23),
+            26 => Ok(Self::AF31),
+            28 => Ok(Self::AF32),
+            30 => Ok(Self::AF33),
+            34 => Ok(Self::AF41),
+            36 => Ok(Self::AF42),
+            38 => Ok(Self::AF43),
+            8 => Ok(Self::CS1),
+            16 => Ok(Self::CS2),
+            24 => Ok(Self::CS3),
+            32 => Ok(Self::CS4),
+            40 => Ok(Self::CS5),
+            48 => Ok(Self::CS6),
+            56 => Ok(Self::CS7),
+            46 => Ok(Self::EF),
+            n if n <= 0x3f => Ok(Self::Other(n)),
+            _ => Err(Error::Other(format!("invalid DSCP value: {value}"))),
```

Comment:
Not sure I read this right, but essentially all 64 possible values for the DSCP field are valid and can happen with real traffic, it is just that some have no symbolic name... that said when receiving say EF it is not guaranteed that the sending networked used that DSCP to signify the EF PHB, with DSCPs essentially anything goes. Still it is is interesting to see if/where DSCP values are changed en route, as some applications attempt to use DSCPs end to end.

Reply:
> Not sure I read this right, but essentially all 64 possible values for the DSCP field are valid and can happen with real traffic, it is just that some have no symbolic name

Exactly right. We'll allow any 6bit value and for those values which are "known" we'll map to a code (which we can display) and the rest will be shown as a number (in hex).